tawazi/profile.py

```diff
@@ -0,0 +1,95 @@
+    # It can be made to use pydantic, you can sort the ExecNodes by their Profile (which takes into consideration the system time).
+    # This might be helpful in infercharge! For example, we can imagine that infercharge imports inferfwk and then reads the profile received from inferserver and cast it (because the profile is just a BaseModel)
+    # and then We can sort the ExecNodes by their consuming time and only display the top 10 or so...
+    #
+    # Best regards,
+    # Bashir
+
+    def __init__(self) -> None:
+        """
+        Profile an execution
+        >>> p1, p2, = Profile(), Profile()
+        >>> p1 == p2
+        True
+        >>> p2.start()
+        >>> p2.finish()
+        >>> p1 < p2
+        True
+        >>> p1 == p2
+        False
+        >>> p1 > p2
+        False
+        >>> l = sorted([p2, p1])
+        >>> l == [p1, p2]
+        True
+        """
+        self.abs_exec_time = 0.0  # [s.] (perf_counter()) system wide time
+        self.process_exec_time = (
+            0.0  # [s.] (process_time()) system & user time consumed by the process
+        )
+        self.thread_exec_time = (
+            0.0  # [s.] (thread_time()) system & user time consumed by the thread
+        )
+
+    def start(self) -> None:
+        self.abs_exec_time = perf_counter()
+        self.process_exec_time = process_time()
+        self.thread_exec_time = thread_time()
+
+    def finish(self) -> None:
+        self.abs_exec_time = perf_counter() - self.abs_exec_time
+        self.process_exec_time = process_time() - self.process_exec_time
+        self.thread_exec_time = thread_time() - self.thread_exec_time
+
+    def __enter__(self) -> "Profile":
+        self.start()
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self.finish()
+
+    def __repr__(self) -> str:
+        return f"{self.abs_exec_time=}\n" f"{self.process_exec_time=}\n" f"{self.thread_exec_time=}"
+
+    def __eq__(self, __o: object) -> bool:
+        if not isinstance(__o, Profile):
+            raise TawaziTypeError(f"{__o} is not an instance of Profile")
+
+        eq = (
+            self.abs_exec_time == __o.abs_exec_time
+            and self.process_exec_time == __o.process_exec_time
+            and self.thread_exec_time == __o.thread_exec_time
+        )
+
+        return eq
+
+    def __lt__(self, __o: object) -> bool:
+        if not isinstance(__o, Profile):
+            raise TawaziTypeError(f"{__o} is not an instance of Profile")
+
+        return self.abs_exec_time < __o.abs_exec_time
+
+
+# This might not be the best way to do this...
+
+# an immutable Profile.
+# Will throw an exception if trying to change it via start/finish
+PROFILE_ZERO = Profile()
+
+
+def error_raiser(msg: str) -> None:
+    raise TawaziBaseException(msg)
+    return None
+
+
+PROFILE_ZERO.start = PROFILE_ZERO.finish = lambda _: error_raiser("PROFILE_ZERO is immutable")  # type: ignore
```

Comment:
hmm not a fan of this pattern

Reply:
(especially since it makes mypy yell)

Reply:
Ok I will replace it with instanciating a new instance of Porfile()...